Draft of /social/process/how standard (#2)
holochain/how: HOW/social/process/how/how-v1.md

---
HOW: /social/process/how
Status: Draft
Created: 2024-11-28

> Version: idx/1

Reviewer: no Source? Could be the same as HOW here : ) Or did I misunderstand what Source was for?

Author: Hah!  Yes there should be a source and it's the self-referential bootstrapping source.  I'll fix that.

---

## Title
HOW Default Process

## Author(s)

zippy <eric@holochain.org>, mattyg <matt@buildyourweb.app>

## Abstract

This HOW describes a default process, format and workflow for how HOWs come into being. HOWs are the primary mechanism for proposing new features, collecting community input on an issue, and documenting decisions affecting Holochain and/or standards and processes used by Holochain hApps or by the Holochain ecosystem.  In other words, this HOW describes **how** the community comes into alignment.

## Motivation

> As Holochain and the Holochain ecosystems evolves, we want a standardized process for proposing, adopting, and documenting standards of all types, including: protocol creation and changes, feature additions, and process improvements, etc.. A well-defined HOW process helps:

Reviewer 2: ```suggestion
As Holochain and the Holochain ecosystems evolves, we want a standardized process for proposing, adopting, and documenting standards of all types, including: protocol creation and changes, feature additions, and process improvements, etc. A well-defined HOW process helps:
```

### HOW Tree

Rather than using the pattern of storing and identifying standards as a long numerical list (as in RFCs or EIPs), HOWs are identified by a semantic path in a hierarchy.  This allows their relationship and types to be more readily understood.  The HOW tree is stored directly as a file hierarchy in a git repo. The HOW tree encompasses different types of standards where the type is represented as the directory structure of in the tree.  The initial structure is:

```
HOW/
├─ info/             // General guidelines or information
├─ social/
│  ├─ process/       // Changes to HOW processes or other organizational/ecosystem processes
│  │  ├─ how/        // This standard
├─ technical/
│  ├─ core/          // Changes to core Holochain protocols
│  ├─ application/   // Ecosystem and Application standards and conventions
```

All HOW standards MUST live as leaves of the tree.  The only files that MAY live at the branches in the tree are a `README.md` file describing the purpose of that part of the tree, and any other files related to that description, i.e. images, diagrams etc.

### HOW Format

> HOW standards live as markdown file named `how.md` at the leaves of this directory structure.  The format of a HOW standard file consists of a markdown header and a markdown body.  The content of the header and body MUST follow a template specified by:

Reviewer: says "named `how.md`", yet this file is `how-v1.md`.

Author: Good point, I didn't fill out a versioning scheme.  Here a couple possiblities:

1. version goes in the file name: `process/how/how-v1.md`
2. version goes in the parent `process/how-v1/how.md/`
3. version goes in the child `process/how/v1/how.md`

In the case where the version is in the directory structure, it's possible that the how itself should allways have a canonical name like `index.md` or `how.md` or `README.md` (because the github will display it automatically when navigating the tree.

Thoughts

Reviewer: I like the way it is now, because it's less overhead to click through in the filesystem / github ui.

Author: Added some clarity about this.

1.  merging the `_requirements.md` files that live at each level of the tree including and above the leaf where the HOW is being added

> 2.  merging in any additional body sections or headers specified by a status change in a process flow template.

Reviewer: Merging is a cute thought, but adds friction to understanding what is needed to contribute. Perhaps the `_requirements.md` should be true templates that live at the leaves, even if that means duplicating a little bit?

Author: Yeah, in the pull request description I mention the idea of having a script that does that merging into a `_templates.md`  file, and I had imagined that that file would live alongside the `requirements` at each step of the hierarchy so that you could see what's required at any point.  I guess I was just worried about "source of truth" but I think it's worth doing this to reduce friction.

Author: fixed

The above format specification allows the templates for parts of the HOW tree to be contextually appropriate for that part of the tree.  For example, `technical/application` standards require a "Reference Implementation" section, what would not be appropriate for a `social/process` or `info` standard.

### HOW Workflow

HOWs transition through three stages: Define, Refine, & Align, each of which may have one or more statuses that formally indicate the steps in the workflow of a HOW for that stage. In the future we intend for there to be many flavors of the stages of this workflow, but this default HOW process specifies just one and the steps for each stage are as follows:

#### Define
This is the ideation stage, when a standard is first thought up and proposed.  The status for this stage is: `Draft`.  The steps in the stage are:

> 1. Author(s) fork the HOW repository, and create a new HOW at the appropriate place in the HOW tree.  Author(s) MUST follow the template for that spot in the tree, as well as add this process (`/social/process/how/how-v1.md`) as the `Source` to the header of the HOW.

Reviewer: I think there should be some guidance for the directory name of the proposal (one or two words, dash separated, gets the point across about what you're proposing, etc).

Author: updated

2. Author(s) MAY discuss ideas in various Holochain community forums.
3. Author(s) MAY get early feedback from core developers.
4. This stage is formally complete when a pull request is made against the HOW repository.  The pull request must be made using the "Create HOW draft" template. Additionally, when submitting the pull request, Author(s) MUST check "allow edits and access to secrets by maintainers".

#### Refine
This is the review stage, when Author(s) are ready for feedback from the core developers and the community at large. The status for this stage are: `REVIEW`, `LAST CALL`.  The steps in this stage are:

> 1. A member of the core team or requested working group will do an initial evaluation of the PR.  If the a team member of working group member is willing to take on stewardship of this HOW they will update the header status to `REVIEW` and will add the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REVIEW`.  If no requested steward is willing take on the HOW it MUST be updated to status `REJECTED` and the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REJECTED` added.

Reviewer: ```suggestion
1. A member of the core team or requested working group will do an initial evaluation of the PR.  If a team member or working group member is willing to take on stewardship of this HOW they will update the header status to `REVIEW` and will add the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REVIEW`.  If no requested steward is willing take on the HOW it MUST be updated to status `REJECTED` and the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REJECTED` added.
```

> 1. A member of the core team or requested working group will do an initial evaluation of the PR.  If the a team member of working group member is willing to take on stewardship of this HOW they will update the header status to `REVIEW` and will add the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REVIEW`.  If no requested steward is willing take on the HOW it MUST be updated to status `REJECTED` and the section(s) specified in `/social/process/how/_requirements.md#Process Template Section: REJECTED` added.

Reviewer: I don't understand the point of setting it to rejected. I assume we don't want to merge every spammy pr that may be opened filling our directory structure with rejections. Shouldn't we just close the PR?

Author: So, this certainly makes sense for spam and other things that are cursorily rejected because they don't actually get taken up by the team, or a working-group or some steward to consider.  But if there were taken up and considered at length, and especially if there's some formal process to say "nope we actually don't want to do this", it seems like it's worth holding on to that information inside the HOW process not just in github close PR comment.

But maybe we could have that also fall under the `ABANDONDED` status with a `Reason` that would be "Formally rejected" for that case?

> 2. Community members make comments and suggested changes on the pull request.

Reviewer 3: I'm a bit concerned that having the discussion directly on the PR makes it really unwieldy to follow if multiple topics need to be discussed in parallel as it is just a linear stream of comments.

Reviewer 2: I personally feel that PR comment threads are a good UX for supporting multiple parallel conversations. This comment for example is parallel discussion to other comment threads above and I get notifications if I participated in a thread. Do you have some idea for alternatives?

3. Author(s) address feedback, making changes as desired, and keeping all commits in the git history without squashing or force-pushing.
4. When most concerns appear to be addressed, the steward makes a call and changes the status to `LAST CALL` at that point the community and Author(s) have 14 days to raise and address any final concerns.
5. Author(s) MAY chose to set the status to `WITHDRAWN` at any time during the Define and Refine stages, at which point the process terminates.

#### Align
This is the concluding stage where either the core team or the working group decides on the HOW.   The statuses for this stage are: `REJECTED`, `LIVE`, `DEFERRED`.  The steps in this stage are:

> 1. What ever acceptance process is use by core team or working group is used (voting, consensus, coin-toss, etc) is used.

Reviewer 3: ```suggestion
1. Whatever acceptance process is use by core team or working group is used (voting, consensus, coin-toss, etc) is used.
```

2. Given the outcome of the process the status of the HOW is updated to `REJECTED`, `DEFERRED` or `LIVE` as appropriate along with addition of sections as per `/social/process/how/_requirements.md`
3. In the case of a HOW being accepted, the PR is merged into the repository, keeping the original git history without squashing.

```mermaid
stateDiagram
   direction LR

    state Define {
        DRAFT
        DRAFT --> WITHDRAWN
        DRAFT --> REVIEW
        WITHDRAWN
    }

    state Refine {
        REVIEW
        REVIEW --> LAST_CALL
        REVIEW --> WITHDRAWN
        LAST_CALL
        LAST_CALL --> REJECTED
        LAST_CALL --> LIVE
        LAST_CALL --> DEFERRED
        LAST_CALL --> WITHDRAWN
    }

    state Align {
        REJECTED

>         ALIVE

Reviewer 3: Does `ALIVE` also include actual implementation for holochain core standards? For app-level standards I guess `ALIVE` is if we agreed on it but for holochain core standards it's not really alive until it's actually implemented in code. But implementation is still considered as part of the `ALIGN` stage maybe? And what about different versions of holochain? A core standard may be implemented in Holochain version 2.0 but not in version 1.0...

Reviewer 2: Hmm that's a good catch. We should track that a core proposal has actually been implemented and at what holochain version.

For now though, since we're starting with HOW only including app-level tech standards, I'd be fine with tabling it.

Author: On the question of implementation:

My intention was that different types of standards would need different degrees thoroughness in speccing/prototyping/implementation to be considered alive.   For example in the IETF/RFC world I believe some standards aren't accepted until there are at least two separate implementations that prove out the workability of the proposal.   So I was assuming that we would end up choosing different versions of the alignment process for different standards and that would be specified there.

But I understand that `LIVE` makes it sound like the standard is actively in use in some code.   So I think I will propose changing the `LIVE` status to just `ACCEPTED` (which pairs better with `REJECTED` anyways).

As to versions, I don't think the status itself needs to incorporate version numbers, I think it's more like the documentation of the Holochain (or other libraries and tools) would indicate which standards (and versions thereof) they are using.

---
HOW: /social/process/how
Source: /social/process/how/how-v1.md
Status: Draft
Created: 2024-11-28
Version: idx/1
---

## Title
HOW Default Process

## Author(s)

zippy <eric@holochain.org>, mattyg <matt@buildyourweb.app>

## Abstract

This HOW describes a default process, format and workflow for how HOWs come into being. HOWs are the primary mechanism for proposing new features, collecting community input on an issue, and documenting decisions affecting Holochain and/or standards and processes used by Holochain hApps or by the Holochain ecosystem.  In other words, this HOW describes **how** the community comes into alignment.

## Motivation

> As Holochain and the Holochain ecosystems evolves, we want a standardized process for proposing, adopting, and documenting standards of all types, including: protocol creation and changes, feature additions, and process improvements, etc. A well-defined HOW process helps:

Reviewer 4: ```suggestion
As Holochain and the Holochain ecosystem evolve, we want a standardized process for proposing, adopting, and documenting standards of all types, including: protocol creation and changes, feature additions, and process improvements, etc. A well-defined HOW process helps:
```

- Maintain transparency in decision-making
- Ensure proper technical documentation

> - Foster community participation

Reviewer 5: This concerns me. We really don't have the tooling in place in Holochain to make this a resilient space to host content. If somebody decides to engage negatively or just flood in comments then we have limited options for dealing with it.

I suggest that we need to create a separate document for discussion that describes how we operate How. Not a proposal but just a technical guide for ourselves. Like what we do with always-online nodes and backups.

Reviewer 2: The idea is that HOW is a github-only process for now. When we and holochain are ready it can be migrated to a happ.

Author: Yes, the idea is to be able to start doing the process now using before we have matured the hApp.

- Track the history of protocol changes
- Enable structured discussions about significant changes
- Enable coherent evolution of both Holochain itself, as well as the patterns and processes in use by the wider Holochain ecosystem.

## Rationale

This standard, as the first and the default process for adding HOWs to the HOW tree fully addresses the motivation:
- as this process is implemented in an open github repo it both maintains transparency of the decisions made and encourages public feedback and comments in the process, as well as tracks all changes made over time.

> - the the HOW tree and the formal process for requirements for parts of the tree and adding to it ensures proper technical documentation as well as creates a clear method for evolving the process itself.

Reviewer 4: ```suggestion
- the HOW tree and the formal process for requirements for parts of the tree and adding to it ensures proper technical documentation as well as creates a clear method for evolving the process itself.
```

## Specification

### General
HOW standards are held in a git repo on github, and use a small number of Github's affordances to help with the process.  This backing-store is intended to be replaced by a Holochain app in the future.  The HOW processes and data formats consist of a tree structure that holds all types of standards for the Holochain framework and ecosystem.

### HOW Tree

> Rather than using the pattern of storing and identifying standards as a long numerical list (as in RFCs or EIPs), HOWs are identified by a semantic path in a hierarchy.  This allows their relationship and types to be more readily understood.  The HOW tree is stored directly as a file hierarchy in a git repo. The HOW tree encompasses different types of standards where the type is represented as the directory structure of in the tree.  The initial structure is:

Reviewer 4: ```suggestion
Rather than using the pattern of storing and identifying standards as a long numerical list (as in RFCs or EIPs), HOWs are identified by a semantic path in a hierarchy.  This allows their relationship and types to be more readily understood.  The HOW tree is stored directly as a file hierarchy in a git repo. The HOW tree encompasses different types of standards where the type is represented as the directory structure in the tree.  The initial structure is:
```

#### Align
This is the concluding stage where either the core team or the working group decides on the HOW.   The statuses for this stage are: `ABANDONED`, `ACCEPTED`, `DEFERRED`.  The steps in this stage are:

1. Whatever acceptance process is use by core team or working group is used (voting, consensus, coin-toss, etc) is used.
2. Given the outcome of the process the status of the HOW is updated to `ABANDONED`, `DEFERRED` or `ACCEPTED` as appropriate along with addition of sections as per `/social/process/how/_requirements.md`
3. In the case of a HOW being accepted, the PR is merged into the repository, keeping the original git history without squashing.

```mermaid
stateDiagram
   direction LR

    state Define {
        DRAFT
        DRAFT --> WITHDRAWN
        DRAFT --> REVIEW
        WITHDRAWN
    }

    state Refine {
        REVIEW
        REVIEW --> LAST_CALL
        REVIEW --> WITHDRAWN
        LAST_CALL
        LAST_CALL --> ABANDONED
        LAST_CALL --> ACCEPTED
        LAST_CALL --> DEFERRED
        LAST_CALL --> WITHDRAWN
    }

    state Align {
        ABANDONED

>         AACCEPTED

Reviewer 2: ```suggestion
        ACCEPTED
```

>         AACCEPTED --> SUPERSEDED

Reviewer 2: ```suggestion
        ACCEPTED --> SUPERSEDED
```
